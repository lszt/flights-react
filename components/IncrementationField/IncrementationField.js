import React, { PropTypes, Component } from 'react';
import './IncrementationField.scss';

class IncrementationField extends Component {

  constructor(props) {
    super(props);
    this.state = {
      value: 0,
    };
  }

  render() {
    return (
      <div className="IncrementationField">
        <a className="decrement" onClick={this.decrement.bind(this)}>-</a>
        <span className="value">{this.state.value}</span>
        <a className="increment" onClick={this.increment.bind(this)}>+</a>
      </div>
    );
  }

  decrement() {
    let newValue = this.state.value - 1;
    if (newValue < 0) newValue = 0;
    this.setState({
      value: newValue,
    });
  }

  increment() {
    this.setState({
      value: this.state.value + 1,
    });
  }
}

export default IncrementationField;
