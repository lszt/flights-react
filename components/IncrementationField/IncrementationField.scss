.IncrementationField {
  a {
    cursor: pointer;
    user-select: none;
  }

  .value {
    padding: 1em;
  }
}
