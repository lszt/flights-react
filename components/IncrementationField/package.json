{
  "name": "IncrementationField",
  "version": "0.0.0",
  "private": true,
  "main": "./IncrementationField.js"
}
