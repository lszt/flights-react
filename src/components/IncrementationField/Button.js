import React from 'react';
import styled from 'styled-components';

const Button = styled.button`
  cursor: pointer;
  border: 0;
  background-color: transparent;
  font-size: 1em;
`;

export default Button;
