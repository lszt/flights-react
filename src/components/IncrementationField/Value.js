import React from 'react';
import styled from 'styled-components';

const Value = styled.span`
  padding: 1em;
`;

export default Value;
