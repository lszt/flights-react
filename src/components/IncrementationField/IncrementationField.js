import React, { PropTypes, Component } from 'react';
import './IncrementationField.scss';

class IncrementationField extends Component {

  constructor(props) {
    super(props);
    const value = typeof this.props.value === 'undefined' ? this.props.minValue : this.props.value;
    if (value < this.props.minValue) {
      throw new Error('Given value ' + value + ' is lower than min value ' + this.props.minValue);
    }
    this.state = {
      value,
    };
  }

  render() {
    if (this.props.readOnly === true) {
      return (
        <div className="IncrementationField readonly">
          <span className="value">{this.state.value}</span>
        </div>
      );
    }

    return (
      <div className="IncrementationField">
        <a className="decrement" onClick={this.decrement.bind(this)}>-</a>
        <span className="value">{this.state.value}</span>
        <a className="increment" onClick={this.increment.bind(this)}>+</a>
      </div>
    );
  }

  decrement() {
    let newValue = this.state.value - 1;
    if (newValue < this.props.minValue) newValue = this.props.minValue;
    this.setState({
      value: newValue,
    }, this.fireChangeEvent.bind(this));
  }

  increment() {
    this.setState({
      value: this.state.value + 1,
    }, this.fireChangeEvent.bind(this));
  }

  fireChangeEvent() {
    if (typeof this.props.onChange === 'function') {
      this.props.onChange({
        target: {
          value: this.state.value,
        },
      });
    }
  }
}

IncrementationField.propTypes = {
  value: PropTypes.number,
  minValue: PropTypes.number,
  onChange: PropTypes.func,
  readOnly: PropTypes.bool,
};

IncrementationField.defaultProps = {
  minValue: 0,
};

export default IncrementationField;
