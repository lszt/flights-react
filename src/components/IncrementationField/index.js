import IncrementationField from './IncrementationField';

export default IncrementationField;
