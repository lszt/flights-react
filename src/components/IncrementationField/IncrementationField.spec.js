import React from 'react';
import ReactDOM from 'react-dom';
import TestUtils from 'react-dom/test-utils';
import bro from 'jsdom-test-browser';
import expect from 'expect';
import IncrementationField from './IncrementationField';
import Utils from '../../../test/Utils';

describe('IncrementationField', () => {
  before((done) => { bro.newBrowser(done); });
  before((done) => { bro.jQueryify(done); });

  it('is initialized with value', () => {
    const component = TestUtils.renderIntoDocument(<IncrementationField value={42}/>);
    const element = ReactDOM.findDOMNode(component);

    expect(component.state.value).toBe(42);
    expect(bro.$('span', element).html()).toBe('42');
  });

  it('is initialized with value if value and min value is specified', () => {
    const component = TestUtils.renderIntoDocument(<IncrementationField value={42} minValue={3}/>);
    const element = ReactDOM.findDOMNode(component);

    expect(component.state.value).toBe(42);
    expect(bro.$('span', element).html()).toBe('42');
  });

  it('throws error if specified value is lower than min value', () => {
    expect(() => TestUtils.renderIntoDocument(<IncrementationField value={2} minValue={3}/>))
      .toThrow('Given value 2 is lower than min value 3');
  });

  it('is initialized with 0 if no value and min value is specified', () => {
    const component = TestUtils.renderIntoDocument(<IncrementationField/>);
    const element = ReactDOM.findDOMNode(component);

    expect(component.state.value).toBe(0);
    expect(bro.$('span', element).html()).toBe('0');
  });

  it('is initialized with min value if no value is specified', () => {
    const component = TestUtils.renderIntoDocument(<IncrementationField minValue={3}/>);
    const element = ReactDOM.findDOMNode(component);

    expect(component.state.value).toBe(3);
    expect(bro.$('span', element).html()).toBe('3');
  });

  it('is incremented by 1 on increment button click', () => {
    const component = TestUtils.renderIntoDocument(<IncrementationField/>);
    const element = ReactDOM.findDOMNode(component);
    const button = bro.$('button:last-child', element).get(0);

    TestUtils.Simulate.click(button);

    expect(component.state.value).toBe(1);
    expect(bro.$('span', element).html()).toBe('1');
  });

  it('is decremented by 1 on decrement button click', () => {
    const component = TestUtils.renderIntoDocument(<IncrementationField value={42}/>);
    const element = ReactDOM.findDOMNode(component);
    const button = bro.$('button:first-child', element).get(0);

    TestUtils.Simulate.click(button);

    expect(component.state.value).toBe(41);
    expect(bro.$('span', element).html()).toBe('41');
  });

  it('cannot have value lower than min value', () => {
    const component = TestUtils.renderIntoDocument(<IncrementationField minValue={3}/>);
    const element = ReactDOM.findDOMNode(component);
    const button = bro.$('button:first-child', element).get(0);

    TestUtils.Simulate.click(button);

    expect(component.state.value).toBe(3);
    expect(bro.$('span', element).html()).toBe('3');
  });

  it('calls onChange handler on increment', () => {
    const handler = Utils.callTracker();
    const component = TestUtils.renderIntoDocument(<IncrementationField value={42} onChange={handler}/>);
    const element = ReactDOM.findDOMNode(component);
    const button = bro.$('button:last-child', element).get(0);

    TestUtils.Simulate.click(button);

    const calls = handler.calls();
    expect(calls.length).toBe(1);
    expect(calls[0][0].target.value).toBe(43);
  });

  it('calls onChange handler on decrement', () => {
    const handler = Utils.callTracker();
    const component = TestUtils.renderIntoDocument(<IncrementationField value={42} onChange={handler}/>);
    const element = ReactDOM.findDOMNode(component);
    const button = bro.$('button:first-child', element).get(0);

    TestUtils.Simulate.click(button);

    const calls = handler.calls();
    expect(calls.length).toBe(1);
    expect(calls[0][0].target.value).toBe(41);
  });
});
