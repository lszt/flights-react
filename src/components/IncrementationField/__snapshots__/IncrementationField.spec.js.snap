// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components IncrementationField cannot have value lower than min value 1`] = `
"<div>
  <styled.button type=\\"button\\" onClick={[Function: bound decrement]} data-cy=\\"undefined-decrement\\">
    -
  </styled.button>
  <styled.span>
    3
  </styled.span>
  <styled.button type=\\"button\\" onClick={[Function: bound increment]} data-cy=\\"undefined-increment\\">
    +
  </styled.button>
</div>"
`;

exports[`components IncrementationField cannot have value lower than min value 2`] = `
"<div>
  <styled.button type=\\"button\\" onClick={[Function: bound decrement]} data-cy=\\"undefined-decrement\\">
    -
  </styled.button>
  <styled.span>
    3
  </styled.span>
  <styled.button type=\\"button\\" onClick={[Function: bound increment]} data-cy=\\"undefined-increment\\">
    +
  </styled.button>
</div>"
`;

exports[`components IncrementationField is decremented by 1 on decrement button click 1`] = `
"<div>
  <styled.button type=\\"button\\" onClick={[Function: bound decrement]} data-cy=\\"undefined-decrement\\">
    -
  </styled.button>
  <styled.span>
    42
  </styled.span>
  <styled.button type=\\"button\\" onClick={[Function: bound increment]} data-cy=\\"undefined-increment\\">
    +
  </styled.button>
</div>"
`;

exports[`components IncrementationField is decremented by 1 on decrement button click 2`] = `
"<div>
  <styled.button type=\\"button\\" onClick={[Function: bound decrement]} data-cy=\\"undefined-decrement\\">
    -
  </styled.button>
  <styled.span>
    41
  </styled.span>
  <styled.button type=\\"button\\" onClick={[Function: bound increment]} data-cy=\\"undefined-increment\\">
    +
  </styled.button>
</div>"
`;

exports[`components IncrementationField is incremented by 1 on increment button click 1`] = `
"<div>
  <styled.button type=\\"button\\" onClick={[Function: bound decrement]} data-cy=\\"undefined-decrement\\">
    -
  </styled.button>
  <styled.span>
    0
  </styled.span>
  <styled.button type=\\"button\\" onClick={[Function: bound increment]} data-cy=\\"undefined-increment\\">
    +
  </styled.button>
</div>"
`;

exports[`components IncrementationField is incremented by 1 on increment button click 2`] = `
"<div>
  <styled.button type=\\"button\\" onClick={[Function: bound decrement]} data-cy=\\"undefined-decrement\\">
    -
  </styled.button>
  <styled.span>
    1
  </styled.span>
  <styled.button type=\\"button\\" onClick={[Function: bound increment]} data-cy=\\"undefined-increment\\">
    +
  </styled.button>
</div>"
`;

exports[`components IncrementationField is initialized with 0 if no value and min value is specified 1`] = `
"<div>
  <styled.button type=\\"button\\" onClick={[Function: bound decrement]} data-cy=\\"undefined-decrement\\">
    -
  </styled.button>
  <styled.span>
    0
  </styled.span>
  <styled.button type=\\"button\\" onClick={[Function: bound increment]} data-cy=\\"undefined-increment\\">
    +
  </styled.button>
</div>"
`;

exports[`components IncrementationField is initialized with min value if no value is specified 1`] = `
"<div>
  <styled.button type=\\"button\\" onClick={[Function: bound decrement]} data-cy=\\"undefined-decrement\\">
    -
  </styled.button>
  <styled.span>
    3
  </styled.span>
  <styled.button type=\\"button\\" onClick={[Function: bound increment]} data-cy=\\"undefined-increment\\">
    +
  </styled.button>
</div>"
`;

exports[`components IncrementationField is initialized with value 1`] = `
"<div>
  <styled.button type=\\"button\\" onClick={[Function: bound decrement]} data-cy=\\"undefined-decrement\\">
    -
  </styled.button>
  <styled.span>
    42
  </styled.span>
  <styled.button type=\\"button\\" onClick={[Function: bound increment]} data-cy=\\"undefined-increment\\">
    +
  </styled.button>
</div>"
`;

exports[`components IncrementationField is initialized with value if value and min value is specified 1`] = `
"<div>
  <styled.button type=\\"button\\" onClick={[Function: bound decrement]} data-cy=\\"undefined-decrement\\">
    -
  </styled.button>
  <styled.span>
    42
  </styled.span>
  <styled.button type=\\"button\\" onClick={[Function: bound increment]} data-cy=\\"undefined-increment\\">
    +
  </styled.button>
</div>"
`;
